### DISCUSSION TITLE
What is the real pybricks license?

### DISCUSSION THREAD

**Discussion by amreo** [pybricks/support#277] [Q&A] [answered]
Hi!
In this page https://pybricks.com/downloads/ the first download link contains a zip file with a very scary "End User License Agreement_ENG.pdf" file but I expected that the zip file contains only the .img file.
The website says that pybricks is open source so I expected that the download file is also open source and not distributed under a EULA license.

Why is the file distributed under a EULA license? What is the real pybricks license? I think the download page is misleading.

**Comment by laurensvalk** [comment, marked as answer]
Pybricks is licensed as MIT.

Pybricks for EV3 is built on ev3dev (see ev3dev.org). ev3dev is open source as well, but under different licenses.

LEGO Education has released that combination on their own website, along with that EULA.

Our download page links to that version as well, so that almost everyone uses the same version. This is much easier to support.

But if you're worried about it, just get the latest ev3dev release.

We will probably be changing the website a bit at some point, since we don't really need a download page for the other hubs.

**Comment by dlech** [comment]
Getting into the nitty gritty details, "Pybricks" is a brand name rather than a specific piece of software. All of the code that we wrote ourselves under the Pybricks name is licensed under the MIT license ([as defined by SPDX](https://spdx.org/licenses/MIT.html)).

However, we also use lots of 3rd party software that have their own licenses. And there are a number of different software products that include "Pybricks" software.

- [LEGO Education's EV3 Python](https://education.lego.com/en-us/product-resources/mindstorms-ev3/teacher-resources/python-for-ev3) is what you downloaded that includes the EULA from LEGO. The disk image contains a full Debian-based Linux distribution and has a wide variety of licenses. The zip file also contains these licenses in a text file. `pybricks-micropython` is one of the many pieces of software in the disk image.
- [ev3dev-stretch](https://www.ev3dev.org) is nearly identical to the LEGO image. It does not include the EULA from LEGO (since it is not a LEGO product). The licenses are in `/usr/share/doc/<package>/copyright` on the disk image itself. Again, `pybricks-micropython` is one of the many pieces of software in the disk image.
- [Pybricks Code](https://code.pybricks.com) is a web-based IDE for programming using Pybricks MicroPython. It uses 3rd party open source web technologies. The license details can be viewed in the *About* dialog box of the app.
- [Pybricks MicroPython](https://github.com/pybricks/pybricks-micropython) is also distributed as a firmware for LEGO Powered Up hubs (in addition to the standalone executable for EV3 mentioned above). Each firmware is distributed with a [ReadMe_OSS.txt](https://github.com/pybricks/pybricks-micropython/blob/master/bricks/technichub/ReadMe_OSS.txt) file that includes all of the licenses. This is different for each hub.
- We also have other assorted command line tools and libraries, such as [pybricksdev](https://github.com/pybricks/pybricksdev). These generally don't bundle 3rd party software (although they do rely on 3rd party packages to function). So they are usually distributed solely under the MIT license.

**Reply by m10d** [reply]
> Getting into the nitty gritty details, "Pybricks" is a brand name rather than a specific piece of software. All of the code that we wrote ourselves under the Pybricks name is licensed under the MIT license ([as defined by SPDX](https://spdx.org/licenses/MIT.html)).

hoping you can update your response in light of the more recent (awesome, thank you!)  license / block-coding functionality?  I plan on buying a license but also strongly support open source and want to understand the license.

**Reply by laurensvalk** [reply]
Thanks for asking. Pybricks is almost entirely open source. Only the pybricks-blocks module isn't at the moment.

Making that one (comparatively small) component a "premium" feature helps support the project, including all of its many free / open source components that we are still very actively working on.

We'd rather open source that component too. But if this one closed source addition means we can take the existing open source components much farther than we otherwise could, and make a great building/coding experience, I think that's worth it.

I understand this is not for everyone, so that's why we've made sure you can still use everything else completely free (and FOSS), or even build your own version of the online code editor, just minus the the blocks component.

**Reply by m10d** [reply]
> Making that one (comparatively small) component a "premium" feature helps support the project, including all of its many free / open source components that we are still very actively working on.

Yes, I agree with this sentiment; it's always a balancing act.  I do appreciate you're offering a perpetual - it's much easier to justify, at least in my opinion, than a subscription; sure SaaS has it's place but everyone is so bombarded with those it becomes a turnoff.

I hope it helps make the project sustainable.  My request is that pybricks team share a ~roadmap of sorts that addresses how the open/closed source will evolve; namely how the the open-source side will stay usable / relevant.  But yeah, it's hard to make development sustainable, I get that.

**Reply by ChrisKader** [reply]
> * [Pybricks MicroPython](https://github.com/pybricks/pybricks-micropython) is also distributed as a firmware for LEGO Powered Up hubs (in addition to the standalone executable for EV3 mentioned above). Each firmware is distributed with a [ReadMe_OSS.txt](https://github.com/pybricks/pybricks-micropython/blob/master/bricks/technichub/ReadMe_OSS.txt) file that includes all of the licenses. This is different for each hub.

The `ReadMe_OSS.txt` license files contain errors. Placeholders such as `Copyright (c) <year> <copyright holders>` need to be completed or their intended meanings specified. The file's end lists "Copyright Notices" which might serve as "fillers" for the placeholders, but this is not explicitly stated. While I would consider submitting a PR, I am hesitant to do so for code with licensing issues, and I believe these changes should be made by someone more versed in legal terminology. Below are a few examples of the errors present in the files.

[bricks/primehub/ReadMe_OSS.txt#17](https://github.com/pybricks/pybricks-micropython/blob/master/bricks/primehub/ReadMe_OSS.txt#L17)
[bricks/primehub/ReadMe_OSS.txt#86](https://github.com/pybricks/pybricks-micropython/blob/master/bricks/primehub/ReadMe_OSS.txt#L86)
[bricks/primehub/ReadMe_OSS.txt#300](https://github.com/pybricks/pybricks-micropython/blob/master/bricks/primehub/ReadMe_OSS.txt#L300)


The current license may not be enforceable due to conflicts between the included licenses and the primary MIT License. It is crucial for projects with components under different licenses to clearly delineate which parts are covered by each license.

**Reply by dlech** [reply]
All of the copyright statements from all included source code files are reproduced at the end of the ReadMe_OSS.txt file.

https://github.com/pybricks/pybricks-micropython/blob/4d457a17fc13d42d9b311df99e05dafaeed67862/bricks/primehub/ReadMe_OSS.txt#L394

The ReadMe_OSS.txt files were created with https://www.fossology.org/ which is a tool specifically made for opens source license compliance.

All source code is available if one needs to know specifically which license applies to which source code file. We are not aware of any conflicts due to license incompatibility.
